cpp/include/kvikio/posix_io.hpp

```diff
@@ -33,7 +35,60 @@ inline constexpr std::size_t posix_bounce_buffer_size = 2 << 23;  // 16 MiB
 namespace detail {
 
 /**
- * @brief Class to retain host memory allocations
+ * @brief Singleton class to retrieve a CUDA stream for device-host copying
+ *
+ * Call `StreamsByThread::get` to get the CUDA stream assigned to the current
+ * CUDA context and thread.
+ */
+class StreamsByThread {
+ private:
+  std::map<std::pair<CUcontext, std::thread::id>, CUstream> _streams;
+
+ public:
+  StreamsByThread() = default;
+  ~StreamsByThread() noexcept
+  {
+    for (auto& [_, stream] : _streams) {
+      try {
+        CUDA_DRIVER_TRY(cudaAPI::instance().StreamDestroy(stream));
```

Comment:
Doesn't this segfault when destroying the singleton? We had to let the libcudf stream pool "leak" to avoid this.

Reply:
Not on my workstation or in CI. Maybe it is because we are using a newer CTK. Do you have a link to the libcudf issue?

Reply:
Thanks for checking